src/assisted_test_infra/test_infra/controllers/node_controllers/oci_api_controller.py

```diff
@@ -0,0 +1,471 @@
+from enum import Enum
+from typing import Any, Callable, List, Optional, Tuple
+
+import libvirt
+import oci
+import waiting
+
+from assisted_test_infra.test_infra import BaseClusterConfig
+from assisted_test_infra.test_infra.controllers.node_controllers.disk import Disk
+from assisted_test_infra.test_infra.controllers.node_controllers.node import Node
+from assisted_test_infra.test_infra.controllers.node_controllers.node_controller import NodeController
+from assisted_test_infra.test_infra.helper_classes.config import BaseNodesConfig
+from assisted_test_infra.test_infra.helper_classes.config.base_oci_config import BaseOciConfig
+from service_client import log
+
+
+def random_name(prefix='', length=8):
+    return prefix + ''.join(random.choice(string.ascii_letters) for i in range(length))
+
+
+class OciState(Enum):
+    RUNNING = "RUNNING"
+    STARTING = "STARTING"
+    STOPPING = "STOPPING"
+    STOPPED = "STOPPED"
+    TERMINATING = "TERMINATING"
+    TERMINATED = "TERMINATED"
+    RESETTING = "RESETTING"
+    RESTARTING = "RESTARTING"
+    ACTIVE = "ACTIVE"
+    SUCCEEDED = "SUCCEEDED"
+    CREATED = "CREATED"
+    IN_PROGRESS: "IN_PROGRESS"
+    FAILED = "FAILED"
+    CANCELING = "CANCELING"
+    CANCELED = "CANCELED"
+
+
+class CleanupResource:
+    """Store resource to be destroyed / deleted.
+
+    The cleanup resource called on teardown and stored in a stack of actions.
+    """
+
+    def __init__(self, callback: Callable, *args, **kwargs):
+        self.callback = callback
+        self.args = args
+        self.kwargs = kwargs
+
+    def __call__(self):
+        log.info(f"Cleaning up resource: {self.callback}({self.args}, {self.kwargs})")
+        self.callback(*self.args, **self.kwargs)
+
+
+class OciApiController(NodeController):
+    """Install openshift cluster with oracle nodes using OCI API code.
+
+    It decouples the terraform implementation from the code, used terraform as blackbox provided by oracle.
+    They create terraform templates for customer, we run them as is.
+    Run oci cluster installation as customer, handle resource creation and deletion.
+
+    Steps:
+    1. Create a bucket (object-storage)
+    2. Upload ISO file to the object-storage bucket
+    3. Create pre authenticated - return ISO download link from OCI storage
+    4. Create a stack , declare the provision configuration.
+    5. Upload infrastructure zip file to the stack and set terraform variable
+    6. Start running job creating oci nodes discovered by redhat cluster and return terraform output files
+    7. Once nodes are discovered we configure manifest return in #6
+    """
+
+    _config: BaseOciConfig
+    # cleanup list , runs on teardown in LIFO cleanup
+    _cleanup_resources = []
+
+    def __init__(self, config: BaseNodesConfig, cluster_config: BaseClusterConfig):
+        super().__init__(config, cluster_config)
+        self.cloud_provider = None
+        self.oci_compartment_oicd = self._config.oci_compartment_oicd
```

Comment:
```suggestion
        self._cloud_provider = None
        self._oci_compartment_oicd = self._config.oci_compartment_oicd
```

they are not meant to be public?